lib/modules.nix

```diff
@@ -230,12 +230,20 @@ rec {
      correspond to the definition of 'loc' in 'opt.file'. */
   mergeOptionDecls = loc: opts:
     foldl' (res: opt:
-      if opt.options ? default && res ? default ||
-         opt.options ? example && res ? example ||
-         opt.options ? description && res ? description ||
-         opt.options ? apply && res ? apply ||
-         # Accept to merge options which have identical types.
-         opt.options ? type && res ? type && opt.options.type.name != res.type.name
+      let t  = res.type;
+          t' = opt.options.type;
+          mergedType = t.typeMerge t'.functor;
+          typeSet = if (bothHave "type") && sameType && (mergedType != null)
+                       then { type = mergedType; }
+                       else {};
+          sameType = t.name == t'.name;
+          bothHave = k: opt.options ? ${k} && res ? ${k};
+      in
+      if bothHave "default" ||
+         bothHave "example" ||
+         bothHave "description" ||
+         bothHave "apply" ||
+         bothHave "type" && (! sameType)
```

Comment:
nit:   bothHave "type" && (mergedType != null)

If they do not have the same type, then they should not be mergeable, and we should report this as an error.
Note, sameType compares the name of the top-level type, but not the wrapped type.  Thus "list" == "list", but not "list(string)" != "list(attrs)".  I do not think this would be wise to create extra strings when we can avoid, as we already have&need the functor.